<center><h1>List dan For Loop</h1></center>

---

### 1. List
Ketika bekerja dengan data kita menyimpan data tersebut di dalam memori komputer dan kemudian memanggilnya sewaktu-waktu saat dibutuhkan untuk menjalankan proses. Data tersebut dapat kita simpan di dalam suatu variabel seperti yanng telah kita pelajari pada bagian sebelumnya.

script :
```python
nama_aplikasi = 'Facebook'
harga = 0.0
mata_uang = 'USD'
total_rating = 297328
user_rating = 3.5
```

Berdasarkan script di atas kita menyimpan data ke dalam masing-masing variabel:
- Data nama_aplikasi 'Facebook' adalah tipe string
- Data harga 0.0 adalah float
- Data mata_uang 'USD' adalah string
- Data total_rating 297328 adalah integer
- Data user_rating 3.5 adalah float

Menyimpang data dengan memberikan nama variabel untuk semua data merupakan sebuah proses yang sangat tidak efisien, sehingga dalam bahasa python kita dapat menyimpan data secara lebih efisien dengan menggunakan **lists**. Contoh data list adalah :

In [1]:
baris_1 = ['Facebook', 0.0, 'USD', 297328, 3.5]
print(baris_1)
print(type(baris_1))

['Facebook', 0.0, 'USD', 297328, 3.5]
<class 'list'>


---
Jadi list dapat dibuat dengan menggunakan kurung kotak, dan kemudian di dalamnya diisi dengan data yang ingin disimpan, dimana setiap data dipisahkan oleh tanda koma.

**Instruksi 2.1** :
- Simpanlah data (``'Instagram', 0.0, 'USD', 238478, 4.0``) dalam bentuk list di dalam variabel yang diberi nama ``baris_2``
- Simpanlah data (``Clash of Clans', 0.0, 'USD', 214892, 4.5``) dalam bentuk list di dalam variabel yang diberi nama ``baris_3``
- Tampilkan ``baris_2`` dan ``baris_3`` dengan menggunakan perintah ``print()``

In [1]:
# Tuliskan jawaban disini
baris_2 = ['Instagram', 0.0, 'USD', 238478, 4.0]
baris_3 = ['Clash of Clans', 0.0, 'USD', 214892, 4.5]
print(baris_2)
print(baris_3)

['Instagram', 0.0, 'USD', 238478, 4.0]
['Clash of Clans', 0.0, 'USD', 214892, 4.5]


---
Dalam sebuah list kita dapat menyimpan data dalam berbagai tipe. List bersifat seperti container yang dapat digunakan untuk menyimpan berbagai jenis data. Seperti pada contoh sebelumnya kita telah bisa menyimpan beberapa tipe data ke dalam sebuah list ``baris_1``. Untuk mengetahui berapa banyak data yang kita simpan di dalam sebuah list maka kita dapat menggunakan perintah ``len()``. Sebagai contoh perhatikan code berikut :

In [5]:
baris_1 = ['Facebook', 0.0, 'USD', 2324894, 3.5]
print(len(baris_1))

list_1 = [1,6,9,7,13,30,66]
print(len(list_1))

list_2 = []
print(len(list_2))

5
7
0


---
Jika bekerja dengan data yang sedikit kita dapat menghitung jumlah data di dalam list dengan sangat mudah. Namun jika kita bekerja dengan data yang sangat banyak maka perintah ``len()`` akan sangat membantu untuk mengetahui jumlah data yang tersimpan di dalam list. 

Setiap data yang tersimpan di dalam list memiliki index. Index ini seperti halnya alamat tempat masing-masing data tersimpan. Sehingga kita dapat menggunakan index data untuk memanggil isi data secara spesifik di dalam sebuah list.

|List (baris_1) |  Index   |
|---------------|----------|
| 'Facebook'    |    0     |
|    0.0        |    1     |
|   'USD'       |    2     |
|  2893489      |    3     |
|    3.5        |    4     |


Jika kita ingin memanggil data yang berapa pada index tertentu kita dapat memberikan perintah dengan format ``nama_list[index]``. Sebagai contoh jika kita ingin memanggil data pada urutan pertama kita dapat memasukkan index ``0``:

In [7]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
print(baris_1[0])

Facebook


In [8]:
print(baris_1[0])
print(baris_1[1])
print(baris_1[2])
print(baris_1[3])
print(baris_1[4])

Facebook
0.0
USD
2974676
3.5


---
Dengan kemudahan memanggil data yang tersimpan dalam list maka kita dapat dengan mudah menjalankan suatu perintah operasi dan menyimpan hasil operasi di dalam memori. Conoth :

In [9]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
baris_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]

beda_rating = baris_2[4] - baris_1[4]
rerata_rating = (baris_1[4] + baris_2[4])/2

print(beda_rating)
print(rerata_rating)

1.0
4.0


---
**Instruksi 2.2** :
Bada sel dibawah kita telah diberikan 3 buah list dengan nama ``baris_1``, ``baris_2``, dan ``baris_3``.
- Buatlah variable dengan nama ``rating_1`` dan simpan data urutan ke-4 dari list ``baris_1``
- Buatlah variabel dengan nama ``rating_2`` dan simpan data urutan ke-4 dari list ``baris_2``
- Buatlah variabel dengan nama ``rating_3`` dan simpan data urutan ke-4 dari list ``baris_3``
- Jumlah ketiga variabel ``rating_1``, ``rating_2``, ``rating_3`` dan simpan dalam variabel baru yang diberi nama ``total``
- Bagilah variabel ``total`` dengan bilangan ``3`` untuk mendapatkan nilai reratanya dan simpan dalam variabel baru yang diberi nama ``rerata``

In [2]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
baris_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
baris_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]

# Tulis jawaban disini
rating_1 = baris_1[4]
rating_2 = baris_2[4]
rating_3 = baris_3[4]
total = rating_1 + rating_2 + rating_3
print(total)

12.5


---
Dalam penomoran index dikenal dua sistem penomoran yaitu
- Positif Index : elemen pertama diberi index 0, elemen kedua diberi index 1, dan seterusnya
- Negatif Index : elemen terakhir diberik index -1, elemen kedua terakhir diberi index -2, dan seterusnya

| elemen_list  |Positif_index|Negatif_Index|
|--------------|-------------|-------------|
| 'Facebook'   |     0       |    -5       |
|    0.0       |     1       |    -4       |
|   'USD'      |     2       |    -3       |
|  2974676     |     3       |    -2       |
|    3.5       |     4       |    -1       |


Dalam prakteknya kita akan sangat sering menggunakan positif_index. Namun dalam kondisi tertentu ketika jumlah data dalam list sangat banyak dan kita ingin melihat data pada bagian akhir maka sistem negatif_index akan sangat berguna.

In [11]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
print(baris_1[-1])
print(baris_1[4])

3.5
3.5


---
Jika kita salah memasukkan index karena jumlah index yang dimasukkan jauh lebih besar dari pada maksimal index yang tersedia, maka outputnya akan muncul error seperti berikut

In [13]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
print(baris_1[6])

IndexError: list index out of range

---
**Instruksi 2.3** :
Gunakan sistem negatif index dalam mengerjakan perintah berikut
- Buatlah variabel ``rating_1`` dan isi dengan data pada elemen terakhir dari ``baris_1``
- Buatlah variabel ``rating_2`` dan isi dengan data pada elemen terakhir dari ``baris_2``
- Buatlah variabel ``rating_3`` dan isi dengan data pada elemen terakhir dari ``baris_3``
- Jumlahkan ``rating_1``, ``rating_2``, ``rating_3``, dan simpan dalam variabel baru yang diberi nama ``total_rating``
- Bagi ``total_rating`` dengan bilangan ``3`` dan kemudian simpan dalam variabel baru dengan nama ``rerata_rating``.
- Tampilkan ``rerata_rating`` dengan perintah ``print()``

In [3]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
baris_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
baris_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]

# Tuliskan jawaban disini
rating_1 = baris_1[-1]
rating_2 = baris_2[-1]
rating_3 = baris_3[-1]
total_rating = rating_1 + rating_2 + rating_3
rerata_rating = total_rating/3
print(rerata_rating)

4.166666666666667


Sering kali dalam membuat suatu program kita harus mengambil data dalam jumlah tertentu dari sebuah list dan kemudian menyimpannya dalam sebuah variabel baru, contoh:

```python
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
nama_aplikasi = baris_1[0]
jumlah_rating = baris_1[3]
rating = baris_1[-1]
```

Cara di atas membuat kita harus mendefinisikan banyak variabel dan membuat code kita tidak efisien dan sulit untuk ditelusuri jika terdapat suatu kesalahan. Cara yang lain dapat dilakukan dengan data yang dipilih dari sebuah list ke dalam list yang lain

```python
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
data_rating =[baris_1[0], baris_1[3], baris_1[-1]]
print(data_rating)

output :
    ['Facebook',2974676, 3.5]
```

---
**Instruksi 2.4** :
Diberikan list seperti pada sel dibawah, dimana pada data pada list berisi nama aplikasi, harga, mata uang, jumlah rating, user rating :
- Buatlah variabel dengan nama ``rating_fb`` dan isi dengan list nama aplikasi, jumlah rating dan user rating dari list ``baris_1``
- Buatlah variabel dengan nama ``rating_insta`` dan isi dengan list nama aplikasi, jumlah rating dan user rating dari list ``baris_2``
- Buatlah variabel dengan nama ``rating_pandora`` dan isi dengan list nama aplikasi, jumlah rating dan user rating dari list ``baris_3``
- Hitunglah rerata user rating dari ``rating_fb``, ``rating_insta``, dan ``rating_pandora``.
   
   -- Jumlahkan semua user rating dan bagi dengan total datanya kemudian simpan dalam variabel ``rating_rerata``
   -- Tampilkan ``rating_rerata`` dengan menggunakan perintah ``print()``

In [5]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
baris_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
baris_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
baris_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
baris_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

# Tuliskan jawaban disini
rating_fb = baris_1[3] + baris_1[4]
rating_insta = baris_2[3] + baris_2[4]
rating_pandora = baris_3[3] + baris_3[4]
rerata_rating = (rating_fb + rating_insta + rating_pandora)/3
print(rerata_rating)

2422350.5


---
Dalam membuat seuatu program, kita seringkali berhadapan dengan list dengan jumlah elemen yang banyak dan kita harus memilih elemen list dalam rentang index tertentu. Untuk melakukan hal tersebut kita dapat memilihnya dengan format perintah berikut ``nama_list[index_awal:index_akhir]``, contohnya :

```python
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
data_slice = baris_1[0:3]
print(data_slice)

output :
    ['Facebook', 0.0, 'USD']
```

Jika kita ingin memilih n elemen pertama atau n elemen terakhir dari sebuah list dapat dilakukan dengan perintah berikut
```python
nama_list[:n]
nama_list[-n:]
```

In [17]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
tiga_awal = baris_1[:3]
tiga_akhir = baris_1[-3:]
print(tiga_awal)
print(tiga_akhir)

['Facebook', 0.0, 'USD']
['USD', 2974676, 3.5]


**Instruksi 2.5** :

Diberikan beberapa list seperti pada sel dibawah
- Pilih 4 elemen pertama pada list ``baris_1`` dan simpan dalam variabel dengan nama ``fb_4_awal``
- Pilih 3 elemen terkahir pada list ``baris_1`` dan simpan dalam variabel dengan nama ``fb_3_akhir``
- Dari variabel ``baris_3`` pilih ``['USD`, 2130805]`` dan simpan dalam variabel ``coc_3_4``.

In [10]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
baris_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
baris_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
baris_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
baris_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

# Tulis jawaban disini
fb_4_awal = baris_1[:4]
fb_3_akhir = baris_1[-3:]
coc_3_4 = baris_3[2:4]
print(fb_4_awal)
print(fb_3_akhir)
print(coc_3_4)

['Facebook', 0.0, 'USD', 2974676]
['USD', 2974676, 3.5]
['USD', 2130805]


---
Pada bagian sebelumnya kita menyimpan sebuah list data kedalam satu buah variabel. Jika kita bekerja dengan jumlah data yang sangat banyak memungkinkan untuk memiliki list data yang jumlahnya sangat banyak. Misalnya kita memiliki list data sebanyak 1000 data, berarti kita harus mendifinisikan sebanyak 1000 variabel untuk menyimpan list data tersebut, tentunya cara ini sangat tidak efisien. Untuk membuat code kita menjadi lebih rapi maka kita dapat menyimpan sekumpulan data list kedalam sebuah list, contohnya:

In [21]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
baris_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
baris_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
baris_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
baris_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

dataset = [baris_1, baris_2, baris_3, baris_4, baris_5]
print(dataset)

[['Facebook', 0.0, 'USD', 2974676, 3.5], ['Instagram', 0.0, 'USD', 2161558, 4.5], ['Clash of Clans', 0.0, 'USD', 2130805, 4.5], ['Temple Run', 0.0, 'USD', 1724546, 4.5], ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]]


---
Dari contoh sebelumnya didefinisikan variabel baru ``dataset`` yang juga merupakan list dari data list.
- Untuk memanggil ``baris_1`` kita dapat memberikan perintah ``dataset[0]``.
- Untuk memanggil ``baris_5`` kita dapat memberikan perintah ``dataset[-1]``.
- untuk memanggil ``baris_1`` dan ``baris_2`` kita dapat memberikan perintah ``dataset[:2]``

In [22]:
print(dataset[0])
print(dataset[-1])
print(dataset[:2])

['Facebook', 0.0, 'USD', 2974676, 3.5]
['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]
[['Facebook', 0.0, 'USD', 2974676, 3.5], ['Instagram', 0.0, 'USD', 2161558, 4.5]]


---
Jika kita ingin memanggil elemen data dalam ``dataset`` kita dapat melakukannya dengan cara berikut, misalnya kita ingin mengambil elemen ke 4 dari ``baris_1`` dalam ``dataset`` yaitu nilai ``3.5``

- pertama kita panggil dulu ``baris_1`` dengan perintah ``dataset[0]``, dan kemudian simpan dalam variabel baru yang kita beri nama ``baris_fb``.
- Jika kita print ``baris_fb`` maka akan diperoleh data yang isinya sama dengan ``baris_1``.
- Kita ambil bagian terakhir dari  data dalam ``baris_fb`` dengan perintah ``baris_fb[-1]`` dan simpan dalam variabel ``rating_fb``.
- Print ``rating_fb`` sehingga hasilnya adalag ``3.5``

In [23]:
baris_fb = dataset[0]
rating_fb = baris_fb[-1]
print(rating_fb)

3.5


---
Dari cara di atas kita memperoleh nilai ``3.5`` melalui dua cara yaitu pertama memanggil ``dataset[0]`` dan kemudian menjalankan perintah ``rating_fb[-1]``. Adapun cara yang lebih singkat adalah dengan cara berikut :

In [24]:
print(dataset[0][-1])

3.5


---
**Instruksi 2.6** :
Dalam sel dibawah telah diberikan 5 baris list.
- Gabungkan ke lima list tersebut menjadi sebuah list dan simpan dengan nama variabel ``baris_dataset``
- Hitunglah nilai rerata user rating dari ``baris_dataset`` dengan memilih elemen yang sesuai, dan kemudian simpan dalam variabel ``rerata_rating``.
* Catatan user rating adalah elemen terakhir dari masing-maising list.

In [11]:
baris_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
baris_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
baris_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
baris_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
baris_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

# Tulis jawaban disini
baris_dataset = [baris_1, baris_2, baris_3, baris_4, baris_5]
rerata_rating = (baris_dataset[0][4] + baris_dataset[1][4] + baris_dataset[2][4] + baris_dataset[3][4] + baris_dataset[4][4])/5
print(rerata_rating)

4.2


---
Sebelumnya kita hanya bekerja dengan data yang jumlahnya sedikit. Sebenarnya data yang kita proses sebelumnya adalah potongan dari data yang banyak yang tersimpan dalam file ``AppleStore.csv``. Jika kita buka file tersebut maka disana terdapat 7.197 baris data dan 16 kolom, sehingga secara total ada sebanyak 7.197 x 16 data poin. Jumlah ini sangat banyak dan akan sangat memakan waktu jika dituliskan secara manual satu per satu. Dengan bahasa python kita dapat memanggil data yang tersimpan di dalam memori komputer dan mengubahnya menjadi susunan list untuk dilakukan proses perhitungan. Adapun cara untuk memanggil file adalah seperti berikut ini :

In [1]:
file_open = open('AppleStore.csv', encoding='utf-8')
file_open

<_io.TextIOWrapper name='AppleStore.csv' mode='r' encoding='utf-8'>

---
untuk membuka file kita dapat menggunakan perintah ``open()``. Setelah file dibuka maka outputnya berupa object ``<_io.TextIOWrapper name='AppleStore.csv' mode='r' encoding='utf-8'>``, penjelasan mengenai object akan dijelaskan pada materi yang lain. Sekarang filenya sudah terbuka dengan perintah ``open()``, selanjutnya adalah membaca isi file. Adapun cara untuk membaca isi file adalah dengan memberikan perintah seperti berikut :

In [2]:
file_open = open('AppleStore.csv', encoding='utf-8')

from csv import reader
read_file = reader(file_open)
read_file

---
Pada code diatas terlihat perintah ``from csv import reader``, perintah ini disebut module. Kita akan mempelajari module dalam materi selanjutnya. Secara sederhana perintah ``reader`` bukan perintah yang terdapat dalam native python, sehingga kita perlu mengimportnya kedalam file python. Setelah file dibaca, maka akan dihasilkan object ``<_csv.reader at 0x1aa96af2d60>``. Selanjutnya kita ubah bentuknya menjadi list dengan perintah ``list()``, seperti pada code berikut :

In [3]:
file_open = open('AppleStore.csv', encoding='utf-8')

from csv import reader
read_file = reader(file_open)
apps_data = list(read_file)

print(apps_data[:5])  # print list pada 5 urutan pertama

[['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1'], ['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1'], ['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1'], ['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1']]


---
Untuk mengetahui banyaknya list yang tersimpan dalam variabel ``apps_data``, maka kita dapat menggunakan perintah ``len()``, sebagai berikut

In [5]:
len(apps_data)

7198

---
Dimana terlihat terdapat sebanyak 7198 baris data, dimana dengan 1 baris data pada bagian awal merupakan list nama kolom.

**Instruksi 2.7** :
Buka file ``AppleStore.csv`` dan simpan dalam bentuk list.
- Buka file dengan perintah ``open()``, dan simpan dalam variabel baru dengan nama ``file_open``
- Baca file yang telah dibuka dengan perintah ``reader`` yang diimport dari module ``csv``, dan simpan hasilnya dalam variabel ``read_file``
- Ubahlah ``read_file`` menjadi bentuk list dengan menggunakan perintah ``list()`` dan simpan dalam variabel ``apps_data``
- Print total elemen dalam list ``apps_data`` dengen perintah ``len()``
- Print baris pertama (baris yang menyimpan nama kolom)
- Print baris kedua dan ketiga (coba dengan index rentang / potongan list)

In [12]:
# Tulis jawaban disini
file_open = open('AppleStore.csv', encoding='utf-8')

from csv import reader
read_file = reader(file_open)
apps_data = list(read_file)
print(len(apps_data))
print(apps_data[0])
print(apps_data[1:3])

7198
['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
[['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1'], ['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']]


### 2. For Loop

Jika kita memiliki data dalam jumlah yang sangat banyak seperti data ``AppleStore.csv``, dan kemudian kita ingin menghitung nilai rerata user rating, maka kita tidak mungkin mengerjakannya dengan cara manual seperti yang telah kita kerjakan sebelumnya dengan data yang hanya 5 baris. Dalam ``AppleStore.csv`` terdapat 7198 baris data, dimana untuk menghitung nilai rerata rating dilakukan dengan mengambil user rating dari setiap baris kemudian dijumlahkan dan dibagi dengan total data. Proses mengambil data user rating dari setiap baris ini merupakan sebuah proses yang berulang, sehingga kita dapat mengerjakannya dengan meggunakan perintah for loop. Setiap proses yang berulang dan memiliki pola tertentu dapat dikerjakan dengan menggunakan perintah for loop, contohnya.

In [10]:
ratings = [4.5, 5, 3.8, 4.3, 4, 3]

for elemen in ratings:
    print(elemen)

4.5
5
3.8
4.3
4
3


---
Contoh di atas memanggil elemen penyusun list ``rating`` dan menampilkan datanya satu persatu secara berurutan. Pada iterasi pertama variabel ``elemen`` akan bernilai ``4.5``, iterasi kedua bernilai ``5``, iterasi ketiga ``3.8``, dan seterusnya. Jika kita kembali ke variabel ``dataset``, maka kita akan mengambil elemen dari masing-masing list penyusun ``dataset``.

In [13]:
row_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
row_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
row_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
row_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
row_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

dataset = [row_1, row_2, row_3, row_4, row_5]

for elemen in dataset :
    rating = elemen
    print(rating)

['Facebook', 0.0, 'USD', 2974676, 3.5]
['Instagram', 0.0, 'USD', 2161558, 4.5]
['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
['Temple Run', 0.0, 'USD', 1724546, 4.5]
['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]


---
Jika kita ingin mengambil elemen terakhir dari list penyusun ``dataset`` maka dapat dilakukan dengan perintah berikut :

In [15]:
for elemen in dataset :
    rating = elemen[-1]
    print(rating)

3.5
4.5
4.5
4.5
4.0


**Instruksi 2.8** :
Diberikan variabel list seperti pada sel dibawah 
- Gunakan perintah ``for loop`` dan ``print`` untuk menampilkan total rating (elemen ke 3) dari variabel ``dataset``.

In [13]:
row_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
row_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
row_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
row_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
row_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

dataset = [row_1, row_2, row_3, row_4, row_5]

# Tulis jawaban disini
for elemen in dataset:
    print(elemen[3])

2974676
2161558
2130805
1724546
1126879


---
Teknik yang baru kita pelajari disebut sebagai ``looping``. Dengan teknik looping kita dapat memanipulasi atau melakukan operasi matematik pada masing-masing elemen penyusun list.

In [18]:
data_list = [2,5,6,8,9]

for elemen in data_list:
    print(elemen*2)

4
10
12
16
18


Code diluar loop dapat berinteraksi dengan code di dalam loop, contohnya 

In [21]:
data_list = [1,3,5,7,8]
total = 0

for elemen in data_list :
    total = total + elemen
    print(total)
print(total/len(data_list))

1
4
9
16
24
4.8


Pada code di atas variabel ``total`` di update dalam setiap iterasinya dijumlahkan dengan elemen. Sehingga cara diatas dapat dipakai untuk menghitung total penjumlahan elemen penyusun sebuah list.

**Instruksi 2.9** :
Diberikan list seperti pada sel dibawah :
- Buatlah variabel ``total`` dan isi dengan nilai ``0``
- Iterasikan (looping) pada variabel ``dataset`` (for elemen in dataset)
- Ambil user rating (data urutan ke 5) dan simban dalam variabel ``rating``
- Update variabel ``total`` dengan menjumlahkannya dengan ``rating``
- Diluar loop, bagi nilai rerata rating (yang disimpan dalam ``total``) dengan jumlah data untuk menghitung nilai rerata rating dang simpang dalam variabel ``rerata_rating``

In [14]:
row_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
row_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
row_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
row_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
row_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

dataset = [row_1, row_2, row_3, row_4, row_5]

# Tulis jawaban disini
total = 0
for elemen in dataset :
    total = total + elemen[4]
rerata_rating = total/len(dataset)
print(rerata_rating)

4.2


---
Selanjutnya kita akan menghitung nilai rerata user rating dari data ``AppleStore.csv`` yang tersimpang dalam database komputer kita. Pertama kita buka file ``AppleStore.csv`` dengan menggunakan perintah ``open()`` dan kemudian mengubahnya menjadi bentuk list.

In [23]:
from csv import reader

open_data = open('AppleStore.csv', encoding='utf-8')
read_data = reader(open_data)
data = list(read_data)

print(len(data))
print(data[:5])

7198
[['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1'], ['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1'], ['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1'], ['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1']]


In [24]:
print(data[0])

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [28]:
print(data[0][8])

user_rating


---
Dari hasil keluaran perintah ``len(data)`` terlihat jumlah data sebanyak 7198. Pada baris pertama adalah nama kolom, sehingga total data poinnya adalah sebanyak 7197. Kita akan menghitung nilai rerata dari ``user_rating`` yang mana berada pada index ke 8. Sehingga kita dapat menghitungnya dengan perintah ``for loop`` seperti berikut

In [30]:
total = 0

for elemen in data:
    rating = elemen[8]
    total = total + rating

TypeError: unsupported operand type(s) for +: 'int' and 'str'

Terdapat error yang menjelaskan bahwa data yang dijumlahkan tidak memiliki tipe yang sama. Variabel ``total`` memiliki tipe ``int`` sedangkan ``rating`` memiliki tipe ``str``. Sehingga agar dapat dijumlahkan maka variabel ``rating`` harus dikonversi dulu menjadi tipe ``int`` atau ``float``. Karena variabel ``rating`` memiliki data yang berbentuk desimal maka kita konversi menjadi tipe ``float``.

In [34]:
total = 0

for elemen in data:
    rating = float(elemen[8])
    total = total + rating

ValueError: could not convert string to float: 'user_rating'

---
Sampai disini juga masih terdapat error. Error ini menjelaskan komputer tidak bisa mengkoversi elemen data ``user_rating`` menjadi ``float``. Tentunya ``user_rating`` yang berupa text tidak bisa dikonversi menjadi angka sehingga menghasilkan error. Akan tetapi data dengan elemen ``user_rating`` ini berada pada kolom pertama yang mana merupakan nama kolom yang seharusnya tidak terlibat dalam perhitungan, sehingga kita harus memilih data tanpa mengikutkan data urutan pertama dengan pemilihan index ``data[1:]``.

In [36]:
total = 0

for elemen in data[1:]:
    rating = float(elemen[8])
    total = total + rating
print(total)

25383.5


---
**Instruksi 2.10** :
Hitunglah rerata user rating dari data ``AppleStore.csv``
- Bukalah data ``AppleStore.csv`` dan ubah menjadi bentuk list dan simpan dalam variabel ``data``
- Buatlah variabel ``rating_sum`` dan isi dengan nilai 0
- Looping pada ``data`` dan pastikan tidak mengikutkan baris pertama (nama kolom) dalam looping.
- Pilih variabel ``user_rating`` pada urutan ke 8 dan simpan dalam variabel ``rating``. Jangan lupa konversi tipe data dari ``str`` menjadi ``float``.
- Update variabel ``rating_sum`` dengan menjumlahkannya dengan nilai ``rating``.
- Diluar loop, bagi nilai ``rating_sum`` dengan total data (gunakan perintah ``len(data[1:])``) dan simpan dalam variabel ``rerata_rating``.
- Tampilkan nilai ``rerata_rating`` dengan menggunakan perintah ``print()``.

In [26]:
# Tuliskan jawaban disini
from csv import reader

open_data = open('AppleStore.csv', encoding='utf-8')
read_data = reader(open_data)
data = list(read_data)

rating_sum = 0
for elemen in data[1:]:
    rating = float(elemen[8])
    rating_sum += rating
rerata_rating = rating_sum/len(data[1:])
print(rerata_rating)

3.526955675976101


---
Cara lain untuk menghitung rerata rating adalah dengan cara menyimpan terlebih dahulu nilai rating dalam sebuah list. Untuk menyimpan atau menambahkan data dalam sebuah list kita dapat menggunakan perintah ``append()``. Contoh penggunakan perintah ``append()`` adalah

In [43]:
list_a = [1,2]
list_b = []
list_c = [3,'d', 2]
list_a.append(3)
list_b.append(12)
list_c.append(2.4)
print(list_a)
print(list_b)
print(list_c)

[1, 2, 3]
[12]
[3, 'd', 2, 2.4]


Pada ``list_a`` awalnya telah berisi nilai [``1``,``2``] kemudian perintah `append(3)` menambahkan angka ``3`` kedalam ``list_a``. Begitu juga pada ``list_b`` dan ``list_c`` , dimana nilai `12` dan `2.4` ditambahkan ke masing-masing list. 

Untuk menghitung nilai rerata dapat dilakukan : 
- Mengambil nilai `user_rating` dan kemudian di simpan ke dalam sebuah list kosong.
- Selanjutnya kita gunakan perintah `sum()` untuk menghitung jumlah total dari nilai `user_rating`
- Untuk menghitung nilai reratanya kita bagi hasil dari ``sum()`` dengan total data dengan menggunakan perintah `len()`

Contohnya :

In [44]:
row_1 = ['Facebook', 0.0, 'USD', 2974676, 3.5]
row_2 = ['Instagram', 0.0, 'USD', 2161558, 4.5]
row_3 = ['Clash of Clans', 0.0, 'USD', 2130805, 4.5]
row_4 = ['Temple Run', 0.0, 'USD', 1724546, 4.5]
row_5 = ['Pandora - Music & Radio', 0.0, 'USD', 1126879, 4.0]

dataset = [row_1, row_2, row_3, row_4, row_5]

ratings = []

for elemen in dataset:
    rating = elemen[-1]
    ratings.append(rating)
    
print(ratings)

rerata_rating = sum(ratings)/len(ratings)
print(rerata_rating)

[3.5, 4.5, 4.5, 4.5, 4.0]
4.2


---
**Instruksi 2.11** :
Hitunglah nilai rerata `user_rating` dari data `AppleStore.csv` :
- Buka dan bacalah data ``AppleStore.csv`` dan ubah menjadi bentuk list dan simpan dalam variabel ``data``
- Buatlah list kosong dengan nama ``ratings``
- Looping pada elemen ``data[1:]`` dengan perintah ``for``.
- Ambil nilai ``user_rating`` (data urutan ke 8) dan simpan dalam variabel ``rating``. Jangan lupa dikonversi menjadi tipe float.
- Tambahkan nilai dalam variabel ``rating`` ke dalam list ``ratings`` dengan menggunakan perintah ``append()``
- Diluar hitunglah jumlah nilai ``ratings`` menggunakan perintah ``sum()`` dan bagi nilai tersebut dengan banyaknya data ``ratings`` dengan perintah ``len()`` untuk mendapatkan nilai reratanya dan simpan dalam variabel ``rerata_rating``.
- Tampilkan ``rerata_rating`` dengan perintah ``print()``

In [27]:
# Tulis jawaban disini


# Tuliskan jawaban disini
from csv import reader

open_data = open('AppleStore.csv', encoding='utf-8')
read_data = reader(open_data)
data = list(read_data)

ratings = []
for elemen in data[1:]:
    rating = float(elemen[8])
    ratings.append(rating)
rerata_rating = sum(ratings)/len(data[1:])
print(rerata_rating)

3.526955675976101
